tests/functional/states/test_kubernetes.py

```diff
@@ -0,0 +1,1426 @@
+
+
+# def test_deployment_present_idempotency(kubernetes, deployment, _deployment_spec):
+#     """
+#     Test kubernetes.deployment_present is idempotent
+#     """
+#     ret = kubernetes.deployment_present(
+#         name=deployment["name"],
+#         namespace=deployment["namespace"],
+#         spec=_deployment_spec,
+#         wait=True,
+#     )
+#     assert ret.result is True
+#     assert "already exists" in ret.comment
+#     assert not ret.changes
+
+
+def test_deployment_present_replace(kubernetes, deployment):
+    """
+    Test kubernetes.deployment_present replaces a deployment
+    """
+    new_spec = {
+        "replicas": 3,
+        "selector": {"matchLabels": {"app": "test"}},
+        "template": {
+            "metadata": {"labels": {"app": "test"}},
+            "spec": {
+                "containers": [
+                    {
+                        "name": "nginx",
+                        "image": "nginx:latest",
+                        "ports": [{"containerPort": 80}],
+                    }
+                ]
+            },
+        },
+    }
+
+    ret = kubernetes.deployment_present(
+        name=deployment["name"],
+        namespace=deployment["namespace"],
+        spec=new_spec,
+        wait=True,
+    )
+
+    assert ret.result is True
+    assert ret.changes["spec"]["replicas"] == 3
+
+
+def test_deployment_present_test_mode(kubernetes, deployment, _deployment_spec):
+    """
+    Test kubernetes.deployment_present in test mode
+    """
+    ret = kubernetes.deployment_present(
+        name=deployment["name"],
+        namespace=deployment["namespace"],
+        spec=_deployment_spec,
+        test=True,
+    )
+
+    assert ret.result is None
+    assert not ret.changes
+
+
+@pytest.mark.parametrize("deployment", [False], indirect=True)
+def test_deployment_present_template_context(kubernetes, deployment, deployment_template):
+    """
+    Test kubernetes.deployment_present with template_context
+    """
+    template_context = {
+        "name": deployment["name"],
+        "namespace": deployment["namespace"],
+        "replicas": 2,
+        "app_label": "test",
+        "image": "nginx:latest",
+        "labels": {"app": "test"},
+    }
+
+    ret = kubernetes.deployment_present(
+        name=deployment["name"],
+        namespace=deployment["namespace"],
+        source=deployment_template,
+        template="jinja",
+        template_context=template_context,
+        wait=True,
+    )
+
+    assert ret.result is True
+    assert ret.changes
+
+
+def test_deployment_absent(kubernetes, deployment):
+    """
+    Test kubernetes.deployment_absent deletes a deployment
+    """
+    ret = kubernetes.deployment_absent(
+        name=deployment["name"],
+        namespace=deployment["namespace"],
+        wait=True,
+    )
+
+    assert ret.result is True
+    assert ret.changes["kubernetes.deployment"]["new"] == "absent"
```

Comment:
suggestion: Ensure that the deployment has actually been deleted (or not, in test mode).

This is valid for all similar tests in this file (I'll stop adding this type of comment).